tests/ui/non_copy_const.rs

```diff
@@ -66,29 +74,38 @@ impl<T: Trait2 + Trait<u32>, U: Trait2> Local<T, U> {
     const ASSOC_3: AtomicUsize = AtomicUsize::new(14); //~ ERROR interior mutable
     const COW: Cow<'static, str> = Cow::Borrowed("tuvwxy");
     const T_SELF: T = T::SELF_2;
-    const U_SELF: U = U::SELF_2; //~ ERROR interior mutable
-    const T_ASSOC: T::NonCopyType = T::ASSOC; //~ ERROR interior mutable
+    const U_SELF: U = U::SELF_2;
+    //~^ ERROR interior mutable
+    //~| HELP consider requiring `U` to be `Copy`
+    const T_ASSOC: T::NonCopyType = T::ASSOC;
+    //~^ ERROR interior mutable
+    //~| HELP consider requiring `<T as Trait<u32>>::NonCopyType` to be `Copy`
     const U_ASSOC: U::CopyType = U::ASSOC_2;
 }
 
 fn main() {
-    ATOMIC.store(1, Ordering::SeqCst);
-    assert_eq!(ATOMIC.load(Ordering::SeqCst), 5);
+    ATOMIC.store(1, Ordering::SeqCst); //~ ERROR interior mutability
+    assert_eq!(ATOMIC.load(Ordering::SeqCst), 5); //~ ERROR interior mutability
+
+    ATOMIC_USIZE_INIT.store(2, Ordering::SeqCst); //~ ERROR interior mutability
+    assert_eq!(ATOMIC_USIZE_INIT.load(Ordering::SeqCst), 0); //~ ERROR interior mutability
+
+    let _once = ONCE_INIT; //~ ERROR interior mutability
 
-    CELL.set(2);
-    assert_eq!(CELL.get(), 6);
+    CELL.set(2); //~ ERROR interior mutability
+    assert_eq!(CELL.get(), 6); //~ ERROR interior mutability
 
     assert_eq!(INTEGER, 8);
     assert!(STRING.is_empty());
 
-    let a = ATOMIC;
+    let a = ATOMIC; //~ ERROR interior mutability
```

Comment:
I think this is a valid use case. We should only be linting taking references to such an item (either due to explicit referencing or due to implicit referencing via method calls)

Reply:
Should `ATOMIC.into_inner()` considered as a use?

Reply:
I don't think so, it's not taking a reference, but consuming the thing by value. But I haven't thought about it in depth. Opinions?

Reply:
🤔  So we should only consider it a use if a reference is taken like:

```rust
let a = &ATOMIC;
let a = &mut ATOMIC;
let a = ATOMIC.load(SeqCst);  // adjusted to &T
let a = ATOMIC.get_mut(); // adjusted to &mut T
```

The followings are all non-uses:

```rust
let a = (ATOMIC, 1);
let a = &(ATOMIC, 1);   // needs to move the variable 
let a = &(ATOMIC, 1).0; // needs to move the variable 
let a = &[ATOMIC][0];
```

Not sure how to deal with projections though, I prefer to lint all of them:

```rust
type Atomic2 = ([AtomicUsize; 1], Vec<AtomicUsize>, u8);
const ATOMIC_2: Atomic2 = ([AtomicUsize::new(4)], Vec::new(), 7);

let a = &ATOMIC_2.0;  // a use?
let a = &ATOMIC_2.0[0];  // a use?
let a = &*ATOMIC_2.1; // a use?
let a = &ATOMIC_2.2; // a use?
```

Reply:
Oh we could access MIR from a lint now? That's new to me 🙂. Let me check, thanks.

Reply:
If you have a `tcx` you can do `tcx.maybe_optimized_mir(def_id)` and then throw a visitor at it.